Tutorial: How to use Next.js middleware to bootstrap feature flags (#6217)
PostHog/posthog.com: contents/tutorials/nextjs-bootstrap-flags.md

---
title: How to use Next.js middleware to bootstrap feature flags
date: 2023-07-03
author: ["ian-vanagas"]
showTitle: true
sidebar: Docs
featuredImage: ../images/tutorials/banners/nextjs-analytics.png
tags: ["configuration", "feature flags"]
---

> Next.js has middleware enabling you to run functions between requests and responses. We can use this to bootstrap feature flags on page load and make them available immediately. This is useful for redirecting, showing components without flickering, avoiding layout changes, and more.

Reviewer: Good intro. Short and clear

Im not sure if "without having to make an additional requests" is technically correct, or if theres a better way to phrase it

```suggestion
[Next.js middleware](https://nextjs.org/docs/app/building-your-application/routing/middleware) enables you to run functions between requests and responses. We can use this to [bootstrap feature flags](/docs/feature-flags/bootstrapping-and-local-evaluation) on page load and make them available immediately – without having to make an additional requests. This is useful for redirecting, showing components without flickering, avoiding layout changes, and more.
```

> In this tutorial, we create a basic Next.js app, set up feature flags with PostHog, write middleware to get those feature flags, and bootstrap them in the client 

Reviewer: ```suggestion
In this tutorial, we create a basic Next.js app, set up [feature flags](/docs/feature-flags) with PostHog, write middleware to get those feature flags, and bootstrap them in the client. 
```

```

Next, go into your app and install PostHog.

```bash
cd middle-bootstrap
npm i posthog-js
```

Once done, set up a PostHog provider by creating a new file in the `app` folder named `providers.js`. In this file, import `PostHogProvider`, initialize PostHog with your project API key and instance address (from your [project settings](https://app.posthog.com/project/settings)), and set up the provider component.

```js
// app/providers.js
'use client'
import posthog from 'posthog-js'
import { PostHogProvider } from 'posthog-js/react'

if (typeof window !== 'undefined') {
  posthog.init("<ph_project_api_key>", {
    api_host: "<ph_instance_address>"
  })
}

export default function PHProvider({ children }) {
  return <PostHogProvider client={posthog}>{children}</PostHogProvider>
}
```

After this, import your provider component into `layout.js` and wrap your main app with it.

```js
// app/layout.js
import './globals.css'
import Providers from './providers'

export default function RootLayout({ children }) {
  return (
    <html lang="en">
      <Providers>
        <body>{children}</body>
      </Providers>
    </html>
  )
}
```

## Create and set up a feature flag

> After setting up PostHog, we need a [feature flag](/docs/feature-flags). To create one, go to the feature flag tab in your PostHog instance, and click "New feature flag." Add a key (like `bootstrap-test`), set rollout to 100% of users, and press save.

Reviewer: ```suggestion
After setting up PostHog, we need a [feature flag](/docs/feature-flags). To create one, go to the [feature flag tab](https://app.posthog.com/feature_flags) in your PostHog instance, and click "New feature flag." Add a key (like `bootstrap-test`), set rollout to 100% of users, and press save.
```

  const ph_project_api_key = '<ph_project_api_key>'
  const ph_cookie_key = `ph_${ph_project_api_key}_posthog`
  const cookie = request.cookies.get(ph_cookie_key);

  let distinct_id;
  if (cookie) {
    distinct_id = JSON.parse(cookie.value).distinct_id;
  } else {
    distinct_id= crypto.randomUUID();
  }
//...
```

With the distinct ID, we can make an API request to evaluate flags for the user. This requires making a request to `https://app.posthog.com/decide?v=3` (or `https://eu.posthog.com/decide?v=3`) with the `api_key` and `distinct_id`. 

> **Note:** We use the API because [Vercel edge middleware](https://vercel.com/docs/concepts/functions/edge-middleware) (which optimizes the speed of this request in your app) has a limited number of packages and `posthog-node` isn’t one of them.

```js
//...
const requestOptions = {
  method: 'POST',
  headers: {
    'Content-Type': 'application/json'
  },
  body: JSON.stringify({
    api_key: ph_project_api_key,
    distinct_id: distinct_id
  })
};

const ph_request = await fetch(
	'https://app.posthog.com/decide?v=3', // or eu
	requestOptions
);
const data = await ph_request.json();
//...
```

> With the flag data, we can create an object to use the client-side bootstrapping and then pass this value as a cookie. Altogether this looks like this:

Reviewer: ```suggestion
With the flag data, we can create an object to use the client-side bootstrapping and then pass this value as a cookie. Altogether, this looks like this:
```

```js
// middleware.js
import { NextResponse } from 'next/server'

export async function middleware(request) {

  const ph_project_api_key = '<ph_project_api_key>'
  const ph_cookie_key = `ph_${ph_project_api_key}_posthog`
  const cookie = request.cookies.get(ph_cookie_key);

  let distinct_id;
  if (cookie) {
    distinct_id = JSON.parse(cookie.value).distinct_id;
  } else {
    distinct_id= crypto.randomUUID();
  }
  
  const requestOptions = {
    method: 'POST',
    headers: {
      'Content-Type': 'application/json'
    },
    body: JSON.stringify({
      api_key: ph_project_api_key,
      distinct_id: distinct_id
    })
  };
  
  const ph_request = await fetch(
		'https://app.posthog.com/decide?v=3', // or eu
		requestOptions
	);
  const data = await ph_request.json();

  const bootstrapData = {
    distinctId: distinct_id,
    featureFlags: data.featureFlags
  }

  const response = NextResponse.next()
  response.cookies.set('bootstrapData', JSON.stringify(bootstrapData))

  return response
}

export const config = {
  matcher: ['/((?!api|_next/static|favicon.ico|vercel.svg|next.svg).*)'],
};
```

> Next, we will handle this data in the client and use it to bootstrap flags 

Reviewer: ```suggestion
Next, we will handle this data in the client and use it to bootstrap flags. 
```
